### ISSUE TITLE
I just did the HTTP client exercise & I dont get why my code passed.

### ISSUE THREAD

**Issue by nfatah** [nodeschool/discussions#1720]
http.get() has a setEncoding() method but how come the code still works even though I set it after console.log(data)

### My Code
```js
var http = require('http');
var url = process.argv[2];
http.get(url, function callback(response){
     response.on("data", function(data){
          console.log(data);
     }).setEncoding('utf8');
     
});
```
### Output
Correctly logs Strings rather than the standard Node Buffer objects if I didnt use setEncoding('utf8').
Just wanna understand.  
### Official solution
```js    
var http = require('http')   
  http.get(process.argv[2], function (response) {  
    response.setEncoding('utf8')  
    response.on('data', console.log)  
    response.on('error', console.error)  
  })  
```

**Comment by maxrimue** [comment]
The code you wrote does not get executed in the order it was written, it works a bit different. Let's use the official solution that you provided as an example:

Actually, even this code would work:
```javascript
var http = require('http')   
  http.get(process.argv[2], function (response) {  
    response.on('data', console.log)  
    response.setEncoding('utf8')  
    response.on('error', console.error)  
  })  
```
The data would still get logged correctly. But why?
What the `response.on()` statement does is that it makes run a certain code when the event `data` happens, in this case its `console.log`. If we only had this statement, the data would get logged as a buffer. However, the `.setEncoding()` follows right after the `response.on()` statement, and it actually gets executed before any `console.log` happens. `console.log` happens when the `data` event happens, that is when data is received. The data won't arrive within milliseconds though, and the engine is not waiting for the `data` event, it continues running the other _until_ the `data` event happens, and then it runs its code, which is the `console.log`.

Each line of code inside the callback, so those three lines:
```javascript
response.setEncoding('utf8')  
response.on('data', console.log)  
response.on('error', console.error) 
```
, take about some milliseconds to complete, but fetching data from any server usually takes one or two seconds. So the three lines are run long before data is received.


This is why your code works too:
```javascript
response.on("data", function(data){
          console.log(data);
     }).setEncoding('utf8');
```

   
To demonstrate what the JavaScript Engine does here, let's make a chronological list of what it does:
1. Do `function...` when event `data` happens
2. Run `.setEncoding()` of `response`
.
.
.
214. Data received, run `function...`
(this example might be a bit inaccurate, but this is pretty much what happens)

   
tl;dr:
- `response.on` means running something when something happens, not immediately
- Those three lines of code are long run before any kind of data is received
- When data is received, and `console.log()` is run, the encoding is already set

**Comment by nfatah** [comment]
Thank you very much! With the information about the delay in server response. I am clear on how asynchronous this code is.
